Source: upstream
Upstream: http://sourceware.org/cgi-bin/cvsweb.cgi/src/gold/expression.cc.diff?cvsroot=src&r1=1.17&r2=1.18
Reason: gold links improperly in some cases, in particular the kernel.
        original report: http://sourceware.org/ml/binutils/2010-11/msg00266.html
        fix: http://sourceware.org/ml/binutils/2010-11/msg00339.html
===================================================================
RCS file: /cvs/src/src/gold/expression.cc,v
retrieving revision 1.17
retrieving revision 1.18
diff -u -r1.17 -r1.18
--- src/gold/expression.cc	2010/10/01 15:02:33	1.17
+++ src/gold/expression.cc	2010/11/18 23:25:08	1.18
@@ -399,18 +399,19 @@
     value(const Expression_eval_info* eei)				\
     {									\
       Output_section* left_section;					\
-      uint64_t left_alignment;						\
+      uint64_t left_alignment = 0;					\
       uint64_t left = this->left_value(eei, &left_section,		\
 				       &left_alignment);		\
       Output_section* right_section;					\
-      uint64_t right_alignment;						\
+      uint64_t right_alignment = 0;					\
       uint64_t right = this->right_value(eei, &right_section,		\
 					 &right_alignment);		\
       if (KEEP_RIGHT && left_section == NULL && right_section != NULL)	\
 	{								\
 	  if (eei->result_section_pointer != NULL)			\
 	    *eei->result_section_pointer = right_section;		\
-	  if (eei->result_alignment_pointer != NULL)			\
+	  if (eei->result_alignment_pointer != NULL			\
+	      && right_alignment > *eei->result_alignment_pointer)	\
 	    *eei->result_alignment_pointer = right_alignment;		\
 	}								\
       else if (KEEP_LEFT						\
@@ -419,8 +420,9 @@
 	{								\
 	  if (eei->result_section_pointer != NULL)			\
 	    *eei->result_section_pointer = left_section;		\
-	  if (eei->result_alignment_pointer != NULL)			\
-	    *eei->result_alignment_pointer = right_alignment;		\
+	  if (eei->result_alignment_pointer != NULL			\
+	      && left_alignment > *eei->result_alignment_pointer)	\
+	    *eei->result_alignment_pointer = left_alignment;		\
 	}								\
       else if ((WARN || left_section != right_section)			\
 	       && (left_section != NULL || right_section != NULL)	\
